### PR TITLE
rebase via comment

### PR DESCRIPTION
**Pull request by clee2000**
Fixes #ISSUE_NUMBER

start rebase bot

random testing on my fork, using `on: issue_comment` https://github.com/clee2000/pytorch/pull/6

stealing both code and an identity from mergebot

corresponding pr on test-infra https://github.com/pytorch/test-infra/pull/305 

doesnt do ghstack

should accept "@pytorchbot rebase me", "@pytorchbot rebase this", "@pytorchmergebot rebase this"

### CODE REVIEW

File: .github/scripts/tryrebase.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    if dry_run:
+        push_result = repo._run_git("push", "--dry-run", "-f", remote_url, refspec)
+    else:
+        push_result = repo._run_git("push", "-f", remote_url, refspec)
+    if "Everything up-to-date" in push_result:
+        gh_post_comment(pr.org, pr.project, pr.pr_num,
+                        f"Tried to rebase and push PR #{pr.pr_num}, but it was already up to date", dry_run=dry_run)
+    else:
+        gh_post_comment(pr.org, pr.project, pr.pr_num,
+                        f"Successfully rebased {pr.head_ref()} onto {onto_branch}, please pull locally " +
```
Comment (janeyx99) [inline .github/scripts/tryrebase.py:34]: Could you add more detailed steps for this? For example, actual commands may be helpful for external contributors + so people could easily copy/paste.

File: .github/scripts/tryrebase.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    repo.fetch(branch, branch)
+    repo._run_git("rebase", onto_branch, branch)
+    if dry_run:
+        push_result = repo._run_git("push", "--dry-run", "-f", remote_url, refspec)
+    else:
+        push_result = repo._run_git("push", "-f", remote_url, refspec)
+    if "Everything up-to-date" in push_result:
+        gh_post_comment(pr.org, pr.project, pr.pr_num,
+                        f"Tried to rebase and push PR #{pr.pr_num}, but it was already up to date", dry_run=dry_run)
+    else:
```
Comment (zengk95) [inline .github/scripts/tryrebase.py:32]: What happens if there's a merge conflict?

**Review by zengk95 — COMMENTED** [review]
think this lgtm

File: .github/scripts/tryrebase.py (hunk 2 above)
Comment (clee2000) [reply .github/scripts/tryrebase.py:32]: An exception is thrown and causes a comment like https://github.com/clee2000/pytorch/pull/6#issuecomment-1106777989 to be made

Comment (zengk95) [reply .github/scripts/tryrebase.py:32]: I see. Can you put the error message in a "```"  code block?

File: .github/scripts/tryrebase.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        gh_post_comment(pr.org, pr.project, pr.pr_num,
+                        f"Successfully rebased {pr.head_ref()} onto {onto_branch}, please pull locally " +
+                        f"before adding more changes (for example, via `git checkout {pr.head_ref()} && " +
+                        "git pull --rebase`", dry_run=dry_run)
```
Comment (janeyx99) [inline .github/scripts/tryrebase.py:36]: nit -- missing )?
```suggestion
                        "git pull --rebase`)", dry_run=dry_run)
```

**Review by janeyx99 — COMMENTED** [review]
Oh! I think there is a way you can test this workflow before landing it. Will comment in the code.

File: .github/workflows/tryrebase.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+name: Rebase PR
+
+on:
+  repository_dispatch:
+    types: [try-rebase]
+
```
Comment (janeyx99) [inline .github/workflows/tryrebase.yml:6]: You can add pull_request as a trigger here and create a mock PR to rebase on line 34.

Comment (seemethere) [reply .github/workflows/tryrebase.yml:6]: I think the testing in https://github.com/clee2000/pytorch/pull/6 should cover most of this

Comment (janeyx99) [reply .github/workflows/tryrebase.yml:6]: oh how did i not notice that--that looks very similar yes

**Review by janeyx99 — APPROVED** [review]

**Comment by seemethere** [comment]
@clee2000 is this still a WIP?

**Review by seemethere — APPROVED** [review]

**Comment by clee2000** [comment]
> @clee2000 is this still a WIP?
@seemethere 

It is a WIP in that not everything is finished, but I'd like to merge so I can test it with probot/the test-infra changes (if there's a way to do that without merging first, I'd be down to do that instead)

**Pull request closed without merging**